apps/api/src/domain/services/query-services/base-query.service.ts

```diff
@@ -0,0 +1,127 @@
+
+    abstract fetchManyDomainModels(
+        specification: ISpecification<TDomainModel>
+    ): Promise<ResultOrError<TDomainModel>[]>;
+
+    abstract buildViewModel(
+        domainInstance: TDomainModel,
+        externalState: InMemorySnapshot
+    ): UViewModel;
+
+    public async fetchById(
+        id: unknown,
+        userOptions: Partial<GeneralQueryOptions> = {}
+    ): Promise<ResultOrError<Maybe<ResourceByIdQueryResult<ViewModelWithTags<UViewModel>>>>> {
+        if (!isAggregateId(id)) return new InternalError(`Invalid id: ${id} for a ${this.type}`);
+
+        const { shouldReturnUnpublishedEntities } = {
+            ...getDefaultQueryOptions(),
+            ...userOptions,
+        };
+
+        const externalState = await this.fetchRequiredExternalState();
+
+        const domainModelSearchResult = await this.fetchDomainModelById(id);
+
+        if (isInternalError(domainModelSearchResult)) {
+            // Database document does not satisfy invariants
+            throw domainModelSearchResult;
+        }
+
+        if (isNotFound(domainModelSearchResult)) return NotFound;
+
+        if (!shouldReturnUnpublishedEntities && !domainModelSearchResult.published) return NotFound;
+
+        const viewModel = this.buildViewModel(domainModelSearchResult, externalState);
+
+        const viewModelWithTags = mixTagsIntoViewModel(
+            viewModel,
+            externalState.tags,
+            this.type
+        ) as UViewModel & {
+            tags: TagViewModel[];
+        };
+
+        return {
+            data: viewModelWithTags,
+            actions: this.commandInfoService.getCommandInfo(
+                // TODO Implement CommandWriteContext on all Resource sub-classes
+                domainModelSearchResult as unknown as CommandWriteContext
+            ),
+        };
+    }
+
+    public async fetchMany(
+        specification: ISpecification<TDomainModel>
+    ): Promise<ResourceIndexQueryResult<ViewModelWithTags<UViewModel>>> {
+        const searchResult = await this.fetchManyDomainModels(specification);
+
+        const requiredExternalState = await this.fetchRequiredExternalState();
+
+        const validInstances = searchResult.filter(
+            (result): result is TDomainModel => !isInternalError(result)
+        );
+
+        const data = validInstances.map((instance) => ({
+            data: mixTagsIntoViewModel(
+                this.buildViewModel(instance, requiredExternalState),
+                requiredExternalState.tags,
+                this.type
+            ),
+            actions: this.commandInfoService.getCommandInfo(
+                // TODO Implement CommandWriteContext on all Resource sub-classes
+                instance as unknown as CommandWriteContext
+            ),
+        }));
+
+        return {
+            data,
+            // TODO Get the Ctor here dynamically!
+            actions: this.commandInfoService.getCommandInfo(Song),
```

Comment:
This will become generic to the resource type?

Reply:
Yes- that's what the TODO is about. I better remember to do this!

Reply:
Done.